packages/utils/src/lib/tree/tree.ts

```diff
@@ -0,0 +1,268 @@
+interface ITreeConfig<T> {
+  getChildren: (dataNode: T) => T[] | undefined | null;
+  getId: (dataNode: T) => string;
+  getLevel: (dataNode: T) => number;
+  reverse?: boolean;
+}
+
+export const TREE_SEPERATOR = '.';
+
+export class Tree<T> {
+  private _data: T[];
+
+  getChildren: (dataNode: T) => T[] | undefined | null;
+  getId: (dataNode: T) => string;
+  getLevel: (dataNode: T) => number;
+
+  constructor(
+    initialData: T[],
+    private config: ITreeConfig<T>
+  ) {
+    this.getChildren = config.getChildren;
+    this.getId = config.getId;
+    this.getLevel = config.getLevel;
+
+    this._data = this.sortDeep(initialData);
+  }
+
+  get data(): ReadonlyArray<T> {
+    return this._data;
+  }
+
+  get flattened(): ReadonlyArray<T> {
```

Comment:
Array type using 'ReadonlyArray<T>' is forbidden. Use 'readonly T[]' instead.eslint@typescript-eslint/array-type